Factories (#3722)
codeigniter4/CodeIgniter4: user_guide_src/source/concepts/factories.rst

#########
Factories
#########

.. contents::
    :local:
    :depth: 2

Introduction
============

Like ``Services``, ``Factories`` are an extension of autoloading that helps keep your code
concise yet optimal, without having to pass around object instances between classes. At its

> simplest, ``Factories`` provides a common way to create a class instance and access it from

Reviewer: ```suggestion
simplest, ``Factories`` provide a common way to create a class instance and access it from
```

anywhere. This is a great way to reuse object states and reduce memory load from keeping
multiple instances loaded across your app.

> Anything can be loaded from a Factory, but the best examples are those classes that are used
> to work on or transmit common data. The framework itself makes heavy use of the ``Config``
> Factory to make sure the correct configuration is in use at any time. 

Reviewer: ```suggestion
Anything can be loaded from Factories, but the best examples are those classes that are used
to work on or transmit common data. The framework itself makes heavy use of the ``Factory``
config to make sure the correct configuration is in use at any time. 
```

> Take a look at ``Models`` as an example. You can access the Factory specific to ``Models``
> by using the magic static method of the Factories class, ``Factories::models()``. Because
> of the common path structure for namespaces and folders, ``Factories`` knows to look for
> files and classes within ``Models``, so you can request a model by its shorthand base name::

Reviewer: ```suggestion
Take a look at ``Models`` as an example. You can access the Factory specific to ``Models``
by using the magic static method of the Factories class, ``Factories::models()``. Because
of the common path structure for namespaces and folders, ``Factories`` know where to look for
files and classes within ``Models``, so you can request a model by its shorthand base name::
```

	use CodeIgniter\Config\Factories;

	$users = Factories::models('UserModel');

Or you could also request a specific class::

	$widgets = Factories::models('Some\Namespace\Models\WidgetModel');

Next time you ask for the same class anywhere in your code, ``Factories`` will be sure
you get back the instance as before::

	class SomeOtherClass
	{
		$widgets = Factories::models('WidgetModel');

> 

Reviewer: ```suggestion
	}
```

Factories Behavior
==================

> The default behavior of ``Factories`` might not always work for ever component. For example,
> say your component name and its path do not align, or you need to be sure that only a certain
> type of class instances are returned, or your component classes require additional parameters
> for their constructor. Learn how to use ``Factories`` and its configuration to meet your needs.

Reviewer: ```suggestion
The default behavior of ``Factories`` might not always work for every component. For example,
say your component name and its path do not align, or you need to be sure that only a certain
type of class instance is returned, or your component classes require additional parameters
for their constructor. Learn how to use ``Factories`` and its configuration to meet your needs.
```

> or your component classes require additional parameters for their constructor.

Isn't the additional arguments for constructor handled by the splat operator (`...`)? I mean after the two `array_shift`, what is left of the `$arguments` is unpacked to the constructor.

Author: That's correct. The header section alludes to the Parameters section that covers this. I reworded it a bit so hopefully it is more clear.

constructor, making it easy to configure your shared instance on-the-fly. For example, say
your app uses a separate database for authentication and you want to be sure that any attempts
to access user records always go through that connection::

	$conn  = db_connect('AuthDatabase');
	$users = Factories::models('UserModel', true, $conn);

Now any time the ``UserModel`` is loaded from ``Factories`` it will in fact be returning the
shared instance that uses the alternate database connection.

Factory Configuration
---------------------

What works for one component may not work for all. ``Factories`` supports alternate behavior
at the component level via configurations. A configuration consists of any of the settings,
which will fall back on the default if not provided.

======================= ============== ==================================================================================================================== ===================================================
Key                     Type           Description                                                                                     Default

> ======================= ============== ==================================================================================================================== ===================================================

Reviewer: Python syntax check screams the table is malformed. I think this is because the dividers for `Description` and `Default` are continuous instead of broken. Also, I think the upper and bottom dividers are not of same length.
